File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys
n, m = map(int, sys.stdin.readline().split())
height = list(map(int, sys.stdin.readline().split()))
start = 0
end = max(height)
ans = 0
while start<=end:
    result = 0
    mid = (start+end)//2
    for i in height:
        if mid<i:
            result+=i-mid
    if result==m:
        ans = mid
        break
    elif result>m:
        start = mid+1
        ans = mid
    else:
        end = mid-1
print(ans)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt(); //나무의 수
        int m = scanner.nextInt(); //나무의 길이

        int start = 0;
        int end = 0;

        int[] trees = new int[n];
        for (int i = 0; i < n; i++) {
            trees[i] = scanner.nextInt();
            if (trees[i] > end) end = trees[i]; //나무의 최대 길이 찾기
        }


        int answer = 0;

        while (start <= end) {
            long result = 0;
            int mid = (start + end) / 2;
            for (int i = 0; i < n; i++) {
                if (mid < trees[i]) result += (trees[i]-mid); //절단기의 길이보다 나무의 길이가 더 길면 가져갈 나무길이 더하기
            }
            if (result >= m) { //적어도 m 미터의 나무 가져가야함
                answer = mid; //가능한 절단 높이 중 가장 큰 값 찾기
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        System.out.println(answer);
    }
}
